File: backend/src/api/routes/rag.ts
import { Router, Request, Response } from 'express';
import { RAGService } from '../../services/RAGService';
import { adaptiveRAGService } from '../../services/AdaptiveRAGService';

const router = Router();

// Create knowledge base
router.post('/knowledge-bases', async (req: Request, res: Response) => {
  try {
    const { name, description, domain, embeddingModel, chunkSize, chunkOverlap } = req.body;

    const knowledgeBase = await RAGService.createKnowledgeBase({
      name,
      description,
      domain,
      embeddingModel,
      chunkSize,
      chunkOverlap,
    });

    res.json(knowledgeBase);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create knowledge base' });
  }
});

// Add documents to knowledge base
router.post('/knowledge-bases/:id/documents', async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const { documents } = req.body;

    const addedDocs = await RAGService.addToKnowledgeBase(id, documents);

    res.json({ count: addedDocs.length, documents: addedDocs });
  } catch (error) {
    res.status(500).json({ error: 'Failed to add documents' });
  }
});

// Retrieve context for query
router.post('/knowledge-bases/:id/retrieve', async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const { query, topK, minRelevance, minTrust } = req.body;

    const context = await RAGService.retrieveContext(id, query, {
      topK,
      minRelevance,
      minTrust,
    });

    res.json(context);
  } catch (error) {
    res.status(500).json({ error: 'Failed to retrieve context' });
  }
});

// Build RAG-enhanced prompt
router.post('/build-prompt', async (req: Request, res: Response) => {
  try {
    const { originalPrompt, knowledgeBaseId, query, options } = req.body;

    const context = await RAGService.retrieveContext(knowledgeBaseId, query, options);
    const ragPrompt = RAGService.buildRAGPrompt(originalPrompt, context, options);

    res.json({ ragPrompt, context });
  } catch (error) {
    res.status(500).json({ error: 'Failed to build RAG prompt' });
  }
});

// Update trust scores
router.post('/knowledge-bases/:id/trust-scores', async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const { feedback } = req.body;

    await RAGService.updateTrustScores(id, feedback);

    res.json({ success: true });
  } catch (error) {
    res.status(500).json({ error: 'Failed to update trust scores' });
  }
});

// Chunk text
router.post('/chunk', async (req: Request, res: Response) => {
  try {
    const { text, chunkSize, overlap } = req.body;

    const chunks = RAGService.chunkText(text, { chunkSize, overlap });

    res.json({ chunks, count: chunks.length });
  } catch (error) {
    res.status(500).json({ error: 'Failed to chunk text' });
  }
});

// ==================== Adaptive RAG Endpoints ====================

/**
 * Retrieve adaptive context with dynamic context packing
 * استرجاع سياق تكيفي مع حزم سياق ديناميكية
 */
router.post('/adaptive/retrieve', async (req: Request, res: Response) => {
  try {
    const {
      knowledgeBaseId,
      query,
      maxChunks = 5,
      minRelevance = 0.7,
      minTrust = 0.5,
      enableSummarization = true,
      maxContextLength = 4000,
      includeSourceTrace = true,
    } = req.body;

    if (!knowledgeBaseId || !query) {
      return res.status(400).json({
        error: 'knowledgeBaseId and query are required',
      });
    }

    const result = await adaptiveRAGService.retrieveAdaptiveContext(
      knowledgeBaseId,
      query,
      {
        maxChunks,
        minRelevance,
        minTrust,
        enableSummarization,
        maxContextLength,
        includeSourceTrace,
      }
    );

    res.json(result);
  } catch (error) {
    console.error('Adaptive RAG retrieval error:', error);
    res.status(500).json({
      error: 'Failed to retrieve adaptive context',
      details: error instanceof Error ? error.message : 'Unknown error',
    });
  }
});

/**
 * Build context mask for trusted sources
 * بناء ماسك سياق للمصادر الموثوقة
 */
router.post('/adaptive/context-mask', async (req: Request, res: Response) => {
  try {
    const {
      knowledgeBaseId,
      trustThreshold = 0.5,
      verifiedOnly = false,
      allowedDomains = [],
      blockedDomains = [],
    } = req.body;

    if (!knowledgeBaseId) {
      return res.status(400).json({ error: 'knowledgeBaseId is required' });
    }

    const contextMask = await adaptiveRAGService.buildContextMask(
      knowledgeBaseId,
      {
        trustThreshold,
        verifiedOnly,
        allowedDomains,
        blockedDomains,
      }
    );

    res.json(contextMask);
  } catch (error) {
    console.error('Context mask error:', error);
    res.status(500).json({ error: 'Failed to build context mask' });
  }
});

/**
 * Get RAG session history
 * الحصول على سجل جلسة RAG
 */
router.get('/adaptive/sessions/:sessionId', async (req: Request, res: Response) => {
  try {
    const { sessionId } = req.params;

    const session = await adaptiveRAGService.getSessionHistory(sessionId);

    if (!session) {
      return res.status(404).json({ error: 'Session not found' });
    }

    res.json(session);
  } catch (error) {
    console.error('Session history error:', error);
    res.status(500).json({ error: 'Failed to retrieve session history' });
  }
});

/**
 * Register a trusted source
 * تسجيل مصدر موثوق
 */
router.post('/adaptive/trusted-sources', async (req: Request, res: Response) => {
  try {
    const {
      name,
      domain,
      url,
      sourceType = 'document',
      baseTrustScore = 0.8,
      autoVerify = false,
    } = req.body;

    if (!name) {
      return res.status(400).json({ error: 'name is required' });
    }

    const trustedSource = await adaptiveRAGService.registerTrustedSource({
      name,
      domain,
      url,
      sourceType,
      baseTrustScore,
      autoVerify,
    });

    res.json(trustedSource);
  } catch (error) {
    console.error('Register trusted source error:', error);
    res.status(500).json({
      error: 'Failed to register trusted source',
      details: error instanceof Error ? error.message : 'Unknown error',
    });
  }
});

/**
 * Summarize chunks with confidence indicators
 * تلخيص المقتطفات مع مؤشرات الثقة
 */
router.post('/adaptive/summarize', async (req: Request, res: Response) => {
  try {
    const { chunks, query } = req.body;

    if (!chunks || !Array.isArray(chunks) || chunks.length === 0) {
      return res.status(400).json({ error: 'chunks array is required' });
    }

    if (!query) {
      return res.status(400).json({ error: 'query is required' });
    }

    const enrichedChunks = await adaptiveRAGService.summarizeChunks(chunks, query);

    res.json({
      enrichedChunks,
      count: enrichedChunks.length,
      avgConfidence: enrichedChunks.reduce((sum, c) => sum + c.confidenceScore, 0) / enrichedChunks.length,
    });
  } catch (error) {
    console.error('Summarize chunks error:', error);
    res.status(500).json({ error: 'Failed to summarize chunks' });
  }
});

export default router;

File: backend/python/app/api/routes/rag.py
"""
RAG API Routes

REST API endpoints for RAG operations including:
- Knowledge base management
- Document ingestion
- Semantic retrieval
- Context building
"""

from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from loguru import logger

from ...core.database import get_session
from ...models.rag_models import (
    DocumentIngestRequest,
    BatchIngestRequest,
    RetrieveRequest,
    KnowledgeBaseCreateRequest,
    ContextBuildRequest,
    IngestResponse,
    BatchIngestResponse,
    RetrieveResponse,
    KnowledgeBase,
    BuiltContext,
    DocumentStatus,
)
from ...services.rag_service import RAGService


router = APIRouter(prefix="/rag", tags=["RAG"])

# Initialize service
rag_service = RAGService()


# ============== Knowledge Base Endpoints ==============

@router.post(
    "/knowledge-bases",
    response_model=KnowledgeBase,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new knowledge base",
)
async def create_knowledge_base(
    request: KnowledgeBaseCreateRequest,
    session: AsyncSession = Depends(get_session),
) -> KnowledgeBase:
    """
    Create a new knowledge base for storing documents.

    A knowledge base is a container for related documents that can be
    searched together.
    """
    try:
        kb = await rag_service.create_knowledge_base(request, session)
        logger.info(f"Created knowledge base: {kb.name}")
        return kb
    except Exception as e:
        logger.error(f"Failed to create knowledge base: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get(
    "/knowledge-bases",
    response_model=List[KnowledgeBase],
    summary="List all knowledge bases",
)
async def list_knowledge_bases(
    session: AsyncSession = Depends(get_session),
) -> List[KnowledgeBase]:
    """List all available knowledge bases with their statistics."""
    try:
        return await rag_service.list_knowledge_bases(session)
    except Exception as e:
        logger.error(f"Failed to list knowledge bases: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get(
    "/knowledge-bases/{kb_id}",
    response_model=KnowledgeBase,
    summary="Get a knowledge base by ID",
)
async def get_knowledge_base(
    kb_id: str,
    session: AsyncSession = Depends(get_session),
) -> KnowledgeBase:
    """Get details of a specific knowledge base."""
    kb = await rag_service.get_knowledge_base(kb_id, session)
    if not kb:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Knowledge base not found: {kb_id}",
        )
    return kb


# ============== Document Ingestion Endpoints ==============

@router.post(
    "/ingest",
    response_model=IngestResponse,
    summary="Ingest a document",
)
async def ingest_document(
    request: DocumentIngestRequest,
    session: AsyncSession = Depends(get_session),
) -> IngestResponse:
    """
    Ingest a document into a knowledge base.

    The document will be:
    1. Chunked according to the specified strategy
    2. Embedded using OpenAI embeddings
    3. Stored with vector indexes for fast retrieval

    **Chunking Strategies:**
    - `fixed_size`: Split by approximate token count
    - `sentence`: Split by sentences
    - `paragraph`: Split by paragraphs
    - `semantic`: Smart splitting by section markers
    """
    result = await rag_service.ingest_document(request, session)

    if result.status == DocumentStatus.FAILED:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=result.message,
        )

    return result


@router.post(
    "/ingest/batch",
    response_model=BatchIngestResponse,
    summary="Batch ingest documents",
)
async def batch_ingest(
    request: BatchIngestRequest,
    session: AsyncSession = Depends(get_session),
) -> BatchIngestResponse:
    """
    Ingest multiple documents at once.

    Useful for bulk loading documents into a knowledge base.
    Returns detailed status for each document.
    """
    return await rag_service.batch_ingest(request, session)


# ============== Retrieval Endpoints ==============

@router.post(
    "/retrieve",
    response_model=RetrieveResponse,
    summary="Retrieve relevant documents",
)
async def retrieve_documents(
    request: RetrieveRequest,
    session: AsyncSession = Depends(get_session),
) -> RetrieveResponse:
    """
    Retrieve documents relevant to a query using semantic search.

    **Search Modes:**
    - Vector search: Pure similarity-based ranking
    - Hybrid search: Combines similarity with trust scores

    Results are ordered by relevance/combined score.
    """
    try:
        return await rag_service.retrieve(request, session)
    except Exception as e:
        logger.error(f"Retrieval failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.post(
    "/context",
    response_model=BuiltContext,
    summary="Build context for a prompt",
)
async def build_context(
    request: ContextBuildRequest,
    session: AsyncSession = Depends(get_session),
) -> BuiltContext:
    """
    Build context for a prompt by retrieving and combining relevant documents.

    This is the main entry point for RAG-augmented prompting:
    1. Searches across specified knowledge bases
    2. Ranks documents by relevance and trust
    3. Combines content within token limits
    4. Returns context with source attribution
    """
    try:
        return await rag_service.build_context(request, session)
    except Exception as e:
        logger.error(f"Context building failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


# ============== Trust Management Endpoints ==============

@router.patch(
    "/documents/{document_id}/trust",
    summary="Update document trust score",
)
async def update_trust_score(
    document_id: str,
    trust_score: float,
    session: AsyncSession = Depends(get_session),
) -> dict:
    """Update the trust score for a document (0.0 to 1.0)."""
    if not 0 <= trust_score <= 1:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Trust score must be between 0 and 1",
        )

    success = await rag_service.update_trust_score(document_id, trust_score, session)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Document not found: {document_id}",
        )

    return {"message": "Trust score updated", "document_id": document_id, "trust_score": trust_score}


@router.post(
    "/documents/{document_id}/verify",
    summary="Verify a document",
)
async def verify_document(
    document_id: str,
    session: AsyncSession = Depends(get_session),
) -> dict:
    """Mark a document as verified, boosting its trust score."""
    success = await rag_service.verify_document(document_id, session)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Document not found: {document_id}",
        )

    return {"message": "Document verified", "document_id": document_id}
